propertyestimator/utils/quantities.py

```diff
@@ -0,0 +1,656 @@
+        """Converts a `VanillaEstimatedQuantity` object into an uncertainties
+        `ufloat` representation.
+
+        Parameters
+        ----------
+        estimated_quantity: VanillaEstimatedQuantity
+            The quantity to create the uncertainties object for.
+
+        Returns
+        -------
+        ufloat
+            The ufloat representation of the estimated quantity object.
+        unit.Unit
+            The unit of the values encoded in the ufloat object.
+        """
+
+        value_in_default_unit_system = estimated_quantity.value.in_unit_system(unit.md_unit_system)
+        uncertainty_in_default_unit_system = estimated_quantity.uncertainty.in_unit_system(unit.md_unit_system)
+
+        value_unit = value_in_default_unit_system.unit
+        unitless_value = estimated_quantity.value.value_in_unit(value_unit)
+
+        uncertainty_unit = uncertainty_in_default_unit_system.unit
+        unitless_uncertainty = estimated_quantity.uncertainty.value_in_unit(uncertainty_unit)
+
+        assert value_unit == uncertainty_unit
+
+        return ufloat(unitless_value, unitless_uncertainty), value_unit
+
+
+class DependantValuesException(ValueError):
+    """An exception which is raised when arithmetic operations are applied
+    to two quantities which are not independent."""
+
+    def __init__(self):
+
+        super().__init__('The two quantities came from the same source, and so'
+                         'cannot be treated as independent. The propagation /'
+                         'calculation of uncertainties must be handled by'
+                         'more sophisticated methods than this class employs.')
+
+
+class TaggedEstimatedQuantity(VanillaEstimatedQuantity):
+    """A slightly more sophisticated version of the `EstimatedQuantity` which inherits
+    from the bare minimum `VanillaEstimatedQuantity` implementation, and extends it
+    by tracking the source of the data stored within.
+
+    Notes
+    -----
+    Care must be taken that the source string is standardised. As an example, this
+    design would not work if comparing a quantity whose source is a protocol id, and
+    another whose source is a trajectory file name.
+
+    Examples
+    --------
+    To add two **independent** quantities together:
+
+    >>> from simtk import unit
+    >>>
+    >>> a_value = 5 * unit.angstrom
+    >>> a_uncertainty = 0.03 * unit.angstrom
+    >>>
+    >>> b_value = 10 * unit.angstrom
+    >>> b_uncertainty = 0.04 * unit.angstrom
+    >>>
+    >>> quantity_a = TaggedEstimatedQuantity(a_value, a_uncertainty, 'calc_325262315:npt_production')
+    >>> quantity_b = TaggedEstimatedQuantity(b_value, b_uncertainty, 'calc_987234582:npt_production')
+    >>>
+    >>> quantity_addition = quantity_a + quantity_b
+
+    To subtract one **independent** quantity from another:
+
+    >>> quantity_subtraction = quantity_b - quantity_a
+
+    Attempting to add quantities from the same source (i.e probably correlated)
```

Comment:
Are there other checks that would make sense to consider here, in addition to checking for identical source tags?